opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp

```diff
@@ -613,8 +619,205 @@ class SimulatorFullyImplicitBlackoilEbos
+            }
+            
+            for (int n = 0; n < 3; n++){
+                wellRate += wr[n][w];
+            }
+            if (is_producer) {
+                if (pu.phase_used[Water] && pu.phase_used[Oil]) {
+                    if (wellRate > 0.00001){
+                        //Watercut
+                        wr[5][w] = wr[0][w]/wellRate;
+                    }
+                }
+                if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                    //Gos-OilRatio (GOR)
+                    wr[6][w] = wr[2][w]/wr[1][w];
+                }
+                if (pu.phase_used[Water] && pu.phase_used[Gas]) {
+                    //Water-Gas Ratio (WGR)       
+                    wr[7][w] = wr[0][w]/wr[2][w];
+                }
+            }
+        }
+        return wr;
+    }
+   
+   //Compute and return produced and injected totals from wells per phase
+   std::vector<double> wellTotals(std::vector<std::vector<double>> wr,
+                                  const Wells* wells,
+                                  const Opm::PhaseUsage& pu)
+   {
+        int nw = wells->number_of_wells;
+        std::vector<double> wtotals (9,0.0);        
+        for (int w = 0; w < nw; ++w) { 
+           const bool is_producer = wells->type[w] == PRODUCER;
+           const bool is_injector = wells->type[w] == INJECTOR;
+           if (is_producer) {
+               //Field Oil produced
+               wtotals[0] += wr[1][w];
+               //Field Water produced
+               wtotals[1] += wr[0][w];
+               //Field Gas produced
+               wtotals[2] += wr[2][w];
+           }
+           if (is_injector){
+               //Field Oil injected
+               wtotals[6] += wr[1][w];
+               //Field Water injected
+               wtotals[7] += wr[0][w];
+               //Field Gas injected
+               wtotals[8] += wr[2][w];
+           }
+        }
+        for (int w = 0; w < nw; ++w) {
+           const bool is_producer = wells->type[w] == PRODUCER;
+           if (is_producer) {
+               if (pu.phase_used[Water] && pu.phase_used[Oil]) {
+                   //Field Watercut
+                   wtotals[3] = wtotals[1]/(wtotals[0] + wtotals[1] + wtotals[2]);
+               }
+               if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                   //Field GOR
+                   wtotals[4] = wtotals[2]/wtotals[0];
+               }
+               if (pu.phase_used[Water] && pu.phase_used[Gas]) {
+                   //Field WGR       
+                   wtotals[5] = wtotals[1]/wtotals[2];
+               }
+	   }
+        }
+        return wtotals;
+   }
+   
+   //Output well report to the .PRT file
+   void outputWellReport(std::vector<std::vector<double>> wr,
+             std::vector<double> wtotals,
+             const Wells* wells,
+             WellState& xw,
+             const UnitSystem& units)
+    {           
+        int nw = wells->number_of_wells;
```

Comment:
I agree and think this point is important: since we already gather a lot of data, handling all the serial vs parallel issues, it makes sense to put this functionality in writeTimeStepSerial(), in SimulatorFullyImplicitBlackoilOutput.cpp line 346 an Opm::data::Wells object is extracted, this is really (almost) just a `std::map<std::string, opm::data::Well>` object, which is defined in opm-output:opm/output/data/Wells.hpp. The string key is the well name, and the data::Well objects contain all you need at this point. There are some things like well potentials that are not included, but we should be able to add them when needed.